--- linux-3.10.0-693.el7.centos.plus.x86_64/include/linux/mod_devicetable.h.orig	2017-07-07 07:37:46.000000000 +0800
+++ linux-3.10.0-693.el7.centos.plus.x86_64/include/linux/mod_devicetable.h	2018-01-03 21:52:23.336452272 +0800
@@ -405,6 +405,7 @@
 /*
  * For Hyper-V devices we use the device guid as the id.
  */
+#define vmbus_device_id hv_vmbus_device_id
 struct hv_vmbus_device_id {
 	uuid_le guid;
 	kernel_ulong_t driver_data;	/* Data private to the driver */
@@ -597,6 +598,7 @@
  *
  * identifies mei client device by uuid and name
  */
+#define mei_device_id mei_cl_device_id
 struct mei_cl_device_id {
 	char name[MEI_CL_NAME_SIZE];
 	uuid_le uuid;
